### DISCUSSION TITLE
InterpolatedStringHandler limitation

### DISCUSSION THREAD

**Discussion by iSazonov** [dotnet/csharplang#6508] [Language Ideas]
Let's consider code example from https://khalidabuhakmeh.com/fun-with-custom-interpolated-string-handlers
```c#
using System.Collections;
using System.Runtime.CompilerServices;
using System.Text;

var numbers = Enumerable.Range(0, 10).ToArray();
Numbers.Write1($"Call {numbers:(###) ###-####} For A Good Time!");
Numbers.Write2('x', $"Call {numbers:(xxx) xxx-xxxx} For A Good Time!");

public static class Numbers
{
    public static void WriteSpecific(PlaceholderInterpolatedStringHandler builder)
    {
        WriteCommon('#', builder);
    }

    public static void WriteCommon(
        char placeholder, 
        [InterpolatedStringHandlerArgument("placeholder")] PlaceholderInterpolatedStringHandler builder
    )
    {
        Console.WriteLine(builder.GetFormattedText());
    }

[InterpolatedStringHandler]
public readonly struct PlaceholderInterpolatedStringHandler
{
    private char Placeholder { get; }
    private StringBuilder Builder { get; }

    public PlaceholderInterpolatedStringHandler(int literalLength, int formattedCount, char placeholder = '#')
        => (Placeholder, Builder) = (placeholder, new StringBuilder());

    public void AppendLiteral(string s) => Builder.Append(s);
    internal string GetFormattedText() => Builder.ToString();
    
    public void AppendFormatted(IEnumerable t, string format)
    {
        var enumerator = t.GetEnumerator();
        foreach (var c in format)
        {
            if (c == Placeholder && enumerator.MoveNext())
                Builder.Append(enumerator.Current);
            else
                Builder.Append(c);
        }
    }
}

```
`WriteCommon` method allows to specify a placeholder and this works.
We want to hide implementation details and hide the placeholder - `WriteSpecific` method is specific method based on `WriteCommon` and it hardcode a placeholder... but really this doesn't work. If we change '#' to another char (ex., '$') this doesn't work - default '#' is used since the handler has created before.

So the problem is if we have a common handler we can not define specific methods in way as in the code example. But what if we need many such specific methods?

(Real scenario comes from [PowerShell trace code](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/utils/StructuredTraceSource.cs). There are many specific methods which use a common code. )

Today only way is to define specific handler for every specific method. This is an unfortunate limitation.

I don't know if this scenario has been discussed before. I hope that a solution is possible. 

---

There is another example https://blog.jetbrains.com/dotnet/2022/02/07/improvements-and-optimizations-for-interpolated-strings-a-look-at-new-language-features-in-csharp-10/

Specific method:
```c#
public void LogDebug(
    [InterpolatedStringHandlerArgument("")]
    ref DebugLoggerStringHandler handler)
{
    if (Level <= LogLevel.Debug)
        Log(LogLevel.Debug, handler.ToStringAndClear());
}
```
Here `LogLevel.Debug` is constant which does the method specific and the const should be in handler:
```c#
[InterpolatedStringHandler]
public ref struct DebugLoggerStringHandler
{
    private DefaultInterpolatedStringHandler _builder;
    public DebugLoggerStringHandler(
        int literalLength,
        int formattedCount,
        Logger logger,
        out bool handlerIsValid)
    {
        handlerIsValid = logger.Level <= LogLevel.Debug;
        _builder = new(literalLength, formattedCount);
    }
...
```
If we want to define LogError(), LogInformation() and so on we have to define new handler for every specific method with one difference - replace `LogLevel.Debug` to `LogLevel.Error` and so on.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).
